corehq/apps/reports/standard/deployments.py

```diff
@@ -209,8 +210,33 @@ def user_query(self, pagination=True):
                                        )
         return user_query
 
+    def get_location_columns(self, grouped_ancestor_locs):
+        from corehq.apps.locations.models import LocationType
+        location_types = LocationType.objects.by_domain('icds-cas')
+        all_user_locations = grouped_ancestor_locs.values()
+        all_user_loc_types = {loc.location_type_id for user_locs in all_user_locations for loc in user_locs}
+        required_loc_columns = [loc_type for loc_type in location_types if loc_type.id in all_user_loc_types]
+        return required_loc_columns
+
+    def user_locations(self, ancestors, location_type):
+        location_list = ['---'] * len(location_type)
+        types_id = [type.id for type in location_type]
+        for loc in ancestors:
+            position = types_id.index(loc.location_type_id)
+            location_list[position] = loc.name
+        return location_list
+
     def process_rows(self, users, fmt_for_export=False):
         rows = []
+        users = list(users)
+        if (toggles.ICDS_DASHBOARD_REPORT_FEATURES.enabled(self.request.couch_user.username) and
+                toggles.DASHBOARD_ICDS_REPORT.enabled(self.domain)) and \
+                self.rendered_as in ['export']:
+
+            location_ids = {user['location_id'] for user in users if user['location_id']}
+            grouped_ancestor_locs = SQLLocation.get_bulk_ancestors(self.domain, location_ids)
+            self.required_loc_columns = self.get_location_columns(grouped_ancestor_locs)
```

Comment:
You could pull out all this logic into a memoized `required_loc_columns` method which calls a memoized `grouped_ancestor_locs` method.  That'd allow you to avoid setting a class attribute here, and would keep the logic in `process_rows` smaller.

Reply:
Hey, 'required_loc_columns' is needed in other function also where `users` is not available on which `required_loc_columns` depends on. Therefore I have put it in a instance variable

Reply:
This defines `required_loc_columns` in the middle of a method, though, which is messy (pylint has an `attribute-defined-outside-init` warning, for instance).  Essentially, this line is a hidden dependency required for anything that references `self.required_loc_columns`, so if this wasn't available, it would be difficult to figure out what went wrong (maybe `process_users` was called instead of `process_rows`, for instance).

`users` suffers the same problem - it's initialized in three different places, but passed around directly, which makes the dependency clear.  You could similarly move the definition of users into a `cached_property`.